<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Using Jupyter (Good), Exercises</span></div>

**Exercise 1 (Learning to interrupt)**

In [3]:
while True:
    print('*', end='')

****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

KeyboardInterrupt: 

Methods to interrupt the kernel:
1. Click Kernel > Interrupt Kernel
2. Click the "i" key twice to interrupt the kernel

**Exercise 2 (Resetting & Restarting the Kernal)**

In [1]:
print = 'PRINT'

In [1]:
print('Hello world')

Hello world


**Exercise 3 (Timing Stuff)**

1. Verify that both codes give the same result (HINT: print()).

In [7]:
squares1 = [i**2 for i in range(10)]
squares2 = list()
for i in range(10):
    squares2.append(i**2)
print(squares1, squares2)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81] [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


2. Use %%timeit to measure the execution time for each example.

In [9]:
%%timeit
squares1 = [i**2 for i in range(10)]

7.35 μs ± 54.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [11]:
%%timeit
squares2 = list()
for i in range(10):
    squares2.append(i**2)

8.37 μs ± 318 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


3. Discuss with a friend to understand what each of the numbers in the %%timeit output represents and write down their meanings.

example: 8.37 μs ± 318 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
The test was run 7 times, with each run consisting of 100,000 loops. Each run took on average 8.37 microseconds, and the standard deviation of these runs were 318 nanoseconds.